### ISSUE TITLE
Allow globally defined exclusion pattern

### ISSUE THREAD

**Issue by jcornaz** [detekt/detekt#2581]
## Expected Behavior
I would like to be able to globally edit the exclusion patterns, rather than chaning them one by one for each rules.

## Current Behavior
Currently each rules has its own exclusion pattern. That leads to a lot of repetition in the config files, and it is especially annoying and error prone when it is time to change the patterns as we are forced to go through all of them one by one.

## Context
I introduced new test source sets in a project and I have to tell detekt to ignore that sourceset for a bunch of rules. Which rules? Simple, all the rules that are already excluding the test sources.

But instead of having to change one line to tell detekt what are the patterns of test sources, I have to go through each rules and edit them one by one.

The configuration styled proposed by #775 would have been perfect.

**Comment by BraisGabin** [comment]
I like the current way to exclude because it allows you to tune your config a lot. But I have the same problem as @jcornaz. I think that detekt should support the current exclusion method and the proposed in #775 or similar. My proposal is:

```yml
exclusion-patterns:
- patterns:
    - **/test/**
    - **/androidTest/**
    - **/*.Test.kt
    - **/*.Spec.kt
    - **/*.Spek.kt
  rules:
    - VariableNaming
    - FunctionNaming
- patterns:
    - *DTO.kt
  rules:
    - Rule4
    - Rule8
...
```

**Comment by BraisGabin** [comment]
To give more context I'm going to explain my use case:

Every time that I update detekt I run `--generate-config`. This rewrites ALL my configurations so then I go change by change checking if I prefer the new defautl configuration or my old one. I do this because I like the have all the posible parameters in the configuration file so if I need to tune any rule I don't need to check any documentation.

This was not a big problem because I have just a few changes in my configuration file so the changes to check were few. But two month ago I added a new source route for my tests. And now I need to revert 24 lines each time that I update detekt.

With the proposed format I should just need to revert one removed line.

**Comment by BraisGabin** [comment]
New format proposed in #2851
```yaml
exclusion-patterns:
  - 'dont apply in tests':
    patterns:
      - '**/test/**'
      - '**/androidTest/**'
      - '**/commonTest/**'
    rules:
      - 'comments'
  - 'mappers are an special case too':
    patterns:
      - '**/jvmTest/**'
      - '**/jsTest/**'
      - '**/iosTest/**'
    rules:
      - 'potential-bugs>LateinitUsage'
      - 'style>MagicNumber'
      - 'style>WildcardImport'
```

**Comment by DcortezMeleth** [comment]
Hi!

Any plans on implementing this in a near future? Maybe I can somehow help?

I would also benefit from such feature as we have some automatically generated class which break few rules and we don't won't modify n rules configs to skip it.
In our scenario actually something as simple as ignoring all rules for listed packages would do the job, but I guess we would like to be more flexible than that.

**Comment by cortinico** [comment]
> Any plans on implementing this in a near future? Maybe I can somehow help?

Not really. We had some attempts in the past, but none of them landed.

**Comment by BraisGabin** [comment]
I think that the last comments about this are here: https://github.com/detekt/detekt/pull/3651#discussion_r619723740 so 2.0 have this feature in its core.

If we want to implement this before 2.0, I think that the key should be a `ReportingExtension`. It have access to the configuration and the findings. Baseline uses the same hook to work.

It would not be perfect because it would be nice to avoid the run of that rule in those files but maybe it could help.

You can even give it a try by your own: https://detekt.github.io/detekt/extensions.html and if you find a general solution we could merge it.

**Comment by BraisGabin** [comment]
https://github.com/detekt/detekt/issues/3251#issuecomment-734835795 should do the trick. Maybe we could add this to the documentation. But I think that we can close this for now.
